tests/core/test_core_components.py

```diff
@@ -327,9 +338,127 @@ def test_LayerStructure(
+    argnames="requested,as_bool,raises,expected",
+    argvalues=[
+        pytest.param(
+            "above",
+            True,
+            does_not_raise(),
+            np.array([True] + [False] * 13),
+            id="basic bool",
+        ),
+        pytest.param(
+            "above",
+            False,
+            does_not_raise(),
+            np.array([0]),
+            id="basic int",
+        ),
+        pytest.param(
+            ["above", "subsoil"],
+            True,
+            does_not_raise(),
+            np.array([True] + [False] * 12 + [True]),
+            id="aggregate bool",
+        ),
+        pytest.param(
+            ["above", "subsoil"],
+            False,
+            does_not_raise(),
+            np.array([0, 13]),
+            id="aggregate int",
+        ),
+        pytest.param(
+            "above_canopy",
+            False,
+            pytest.raises(ValueError),
+            None,
+            id="bad string",
+        ),
+        pytest.param(
+            ["above", "lowsoil"],
+            False,
+            pytest.raises(ValueError),
+            None,
+            id="bad list",
+        ),
+    ],
+)
+def test_LayerStructure_get_index(
+    caplog, fixture_core_components, requested, as_bool, raises, expected
+):
+    """Test the get index method."""
+
+    with raises:
+        indices = fixture_core_components.layer_structure.get_indices(
+            requested, as_bool=as_bool
         )
 
+    if isinstance(raises, does_not_raise):
+        assert np.all(np.equal(indices, expected))
+
+
+def test_LayerStructure_set_filled_canopy():
+    """Test the set_filled_canopy_method."""
+
+    from virtual_ecosystem.core.config import Config
+    from virtual_ecosystem.core.core_components import LayerStructure
+
+    cfg = Config(cfg_strings="[core]")
+    layer_structure = LayerStructure(
+        cfg, n_cells=9, max_depth_of_microbial_activity=0.25
+    )
+
+    # Check the defaults
+    assert np.all(
+        np.equal(np.isnan(layer_structure.lowest_canopy_filled), np.repeat(True, 9))
+    )
+    assert np.allclose(
+        layer_structure.get_indices("filled_canopy", as_bool=False), np.array([])
+    )
+
+    # Set an aggregate index
+    agg_roles = ["above", "filled_canopy", "surface"]
+    agg_index = frozenset(agg_roles)
+    agg_values = layer_structure.get_indices(agg_roles, as_bool=False)
+
+    assert agg_index in layer_structure._role_indices_int
+    assert np.all(np.equal(agg_values, np.array([0, 11])))
+
+    # Run the method
+    canopy_heights = np.full(
+        (layer_structure.n_canopy_layers, layer_structure._n_cells), np.nan
+    )
+    canopy_heights[0:8, 0:8] = np.where(np.flipud(np.tri(8)), 1, np.nan)
+
+    layer_structure.set_filled_canopy(canopy_heights=canopy_heights)
+
+    # Check everything has been set
+    assert np.allclose(
+        layer_structure.lowest_canopy_filled,
+        np.concatenate([np.arange(8, 0, -1), [np.nan]]),
+        equal_nan=True,
+    )
```

Comment:
This test is a bit unclear to me

Reply:
Yeah - it does a tonne of work. I've extended the docstring to make it clearer what the test stages are.